### DISCUSSION TITLE
Find references disabled

### DISCUSSION THREAD

**Discussion by terpstra** [scalameta/metals#2304] [Q&A] [answered]
I've been writing a BSP to connect with scala-metals. So far this has been super easy thanks to adequate documentation and `~/.cache/metals/bsp.trace.json`.

However, when I open metals doctor, I see that each project has three green checkmarks for Diagnostics, Goto definition, and Completions, but not for Find references. I used semanticdb to get the first three working. Is there another compiler plugin I need to enable the last feature?

Thanks a lot!

**Comment by tgodzik** [comment, marked as answer]
Thanks for the question! Actually diagnostics, goto definition and completions all work without the semanticdb. Semanticdb is only needed for references, implementations and renames. Whenever references is not found it means that Metals did not detect semanticdb flags within the build information that was given to the Metals server. If you did enable semanticdb, but Metals does not have that information the doctor will report it as an issue.

**Reply by terpstra** [reply]
I'm fairly certain semanticdb is working at least partially, because when I do not include the semanticdb files, I no longer get types+documentation from libraries in my workspace when I mouse-over symbols. Also, renaming variables does work, updating uses in other files.
![Screen Shot 2020-12-14 at 8 31 29 PM](https://user-images.githubusercontent.com/1101706/102171473-7321db00-3e4b-11eb-8d61-5fc10836a6fd.png)

**Reply by tgodzik** [reply]
Rename only works if references works, so most likely semanticdb files are produced and Metals is finding it, but the doctor is not finding semanticdb compiler options in the scalacOptions. So maybe you are just not forwarding that information?

**Reply by terpstra** [reply]
That was the problem! Thanks a lot.
I had indeed forgotten to include plugins in the options reported for buildTarget/scalacOptions. I keep those in a separate property and forgot to include them.

**Reply by terpstra** [reply]
Now it just warns me to `Add the compiler option "-P:semanticdb:sourceroot:/scratch/terpstra" to ensure code navigation works.`.
However, I'm not going to do that as that would make the compile line used depend on the directory in which it is invoked, which would break build relocatability. Everything seems to work, despite that warning. Can I safely ignore it?

**Reply by tgodzik** [reply]
sourceroot is used by semanticdb to create a structure in `META-INF` directory. It might be possible that the default works in your case, so maybe you could ignore it indeed.

**Reply by terpstra** [reply]
Awesome. Thanks again!
